library/core/src/str/validations.rs

```diff
@@ -112,162 +112,290 @@ where
+                        block_loop!(8);
                     }
-                    if next!() as i8 >= -64 {
-                        err!(Some(2))
+
+                    // check 4-word blocks for non-ASCII bytes
+                    while curr < block_end_4x {
+                        block_loop!(4);
                     }
-                }
-                4 => {
-                    match (first, next!()) {
-                        (0xF0, 0x90..=0xBF) | (0xF1..=0xF3, 0x80..=0xBF) | (0xF4, 0x80..=0x8F) => {}
-                        _ => err!(Some(1)),
+
+                    // check 2-word blocks for non-ASCII bytes
+                    while curr < block_end_2x {
+                        block_loop!(2);
                     }
-                    if next!() as i8 >= -64 {
-                        err!(Some(2))
+
+                    // `(size_of::<usize>() * 2) + (align_of::<usize> - 1)`
+                    // bytes remain at most
+                    break None;
+                };
+
+                // if the block loops were stopped due to a non-ascii byte
+                // in some block, do another block-wise search using the last
+                // used block-size for the specific byte in the previous block
+                // in order to skip checking all bytes up to that one
+                // individually.
+                // NOTE: this operation does not auto-vectorize well, so it is
+                // done only in case a non-ASCII byte is actually found
+                if let Some(len) = len {
+                    // SAFETY: `curr` has not changed since the last block loop,
+                    // so it still points at a byte marking the beginning of a
+                    // word-sized block of the given `len`
+                    let block = unsafe {
+                        let ptr = start.add(curr) as *const usize;
+                        core::slice::from_raw_parts(ptr, len)
+                    };
+
+                    // calculate the amount of bytes that can be skipped without
+                    // having to check them individually
+                    let (skip, non_ascii) = non_ascii_byte_position(block);
+                    curr += skip;
+
+                    // if a non-ASCII byte was found, skip the subsequent
+                    // byte-wise loop and go straight back to the main loop
+                    if non_ascii {
+                        continue;
                     }
-                    if next!() as i8 >= -64 {
-                        err!(Some(3))
+                }
+
+                // ...otherwise, fall back to byte-wise checks
+                while curr < end && buf[curr] < 128 {
+                    curr += 1;
+                }
+            } else {
+                // byte is < 128 (ASCII), but pointer is not word-aligned, skip
+                // until the loop reaches the next word-aligned block)
+                for _ in 0..offset {
```

Comment:
This can just use a `while` like the rest of these cases, as we've removed `~const` traits in the rest of std.
```suggestion
                let mut i = 0;
                while i < offset {
                    i += 1;
```

Reply:
Yeah, sorry for missing that one, Fixed in the latest commit.